### PR TITLE
Add BICGSTAB Solver

### PR DESCRIPTION
**Pull request by stefantakacs**
This is a spin-of of the pMultigrid-branch

Co-authored-by: roeltielen <roeltielen@hotmail.com>

### CODE REVIEW

File: src/gsSolver/gsBiCgStab.hpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        return true;
+
+    // m_res = rhs - A * x;
+    m_mat->apply(x,m_tmp);
+    m_res = rhs - m_tmp;
+
+    m_r0 = m_res;
+
+    m_p.resize( m_mat->cols(), 1); m_p.setZero();
+    m_v.resize( m_mat->cols(), 1); m_v.setZero();
```
Comment (filiatra) [inline src/gsSolver/gsBiCgStab.hpp]: Could be shortened as 
m_p.setZero( m_mat->cols(), 1);
m_v.setZero( m_mat->cols(), 1);

File: src/gsSolver/gsBiCgStab.hpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+template<class T>
+bool gsBiCgStab<T>::step( typename gsBiCgStab<T>::VectorType& x )
+{
+    T rho_old = m_rho;
+    m_rho = m_r0.col(0).dot(m_res.col(0));
+
+    if (abs(m_rho) < m_restartThereshold * m_r0.col(0).dot(m_r0.col(0)) )
+    {
+        gsInfo << "Residual too orthogonal, restart with new r0 \n";
```
Comment (filiatra) [inline src/gsSolver/gsBiCgStab.hpp]: maybe improve phrase:
Residual almost orthogonal

File: src/gsSolver/gsBiCgStab.hpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    T beta = (m_rho/rho_old)*(m_alpha/m_w);
+    m_p = m_res + beta*(m_p - m_w * m_v);
+
+    // Apply preconditioning by solving Ahat m_y = m_p
+    m_precond->apply(m_p, m_y);
+    // m_v = A * m_y;
+    m_mat->apply(m_y, m_v);
+    m_alpha = m_rho/(m_r0.col(0).dot(m_v.col(0)));
+
+    m_s = m_res - m_alpha * m_v;
```
Comment (filiatra) [inline src/gsSolver/gsBiCgStab.hpp]: m_s.noalias() = ..

File: src/gsSolver/gsBiCgStab.hpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    m_mat->apply(m_z, m_t);
+
+    if (m_t.col(0).dot(m_t.col(0)) > 0)
+        m_w = m_t.col(0).dot(m_s.col(0))/m_t.col(0).dot(m_t.col(0));
+    else
+        m_w = 0;
+
+    // Update iterate and residual
+    x += m_alpha * m_y + m_w * m_z;
+    m_res -= m_alpha * m_v + m_w * m_t;
```
Comment (filiatra) [inline src/gsSolver/gsBiCgStab.hpp]: x.noalias() += ..
m_res.noalias() -= ..

File: src/gsSolver/gsBiCgStab.hpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+    m_error = m_res.norm() / m_rhs_norm;
+    if (m_error < m_tol)
```
Comment (filiatra) [inline src/gsSolver/gsBiCgStab.hpp]: return (m_error < m_tol);

**Review by filiatra — APPROVED** [review]

**Comment by stefantakacs** [comment]
Included also some minor int->index_t that I previously only had locally.

File: src/gsSolver/gsBiCgStab.hpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+}
+
+template<class T>
+bool gsBiCgStab<T>::step( typename gsBiCgStab<T>::VectorType& x )
+{
+    T rho_old = m_rho;
+    m_rho = m_r0.col(0).dot(m_res.col(0));
+
+    if (abs(m_rho) < m_restartThereshold * m_r0.col(0).dot(m_r0.col(0)) )
```
Comment (filiatra) [inline src/gsSolver/gsBiCgStab.hpp]: math::abs

**Review by filiatra — APPROVED** [review]

**Comment by stefantakacs** [comment]
@filiatra I am wondering why the newest revision has 4 failing tests, the other revisions were just fine. Due to CDash, the problem seems to be independent of this PR.

**Comment by filiatra** [comment]
> @filiatra I am wondering why the newest revision has 4 failing tests, the other revisions were just fine. Due to CDash, the problem seems to be independent of this PR.

Indeed those are independent of this PR, you can merge this one I think

**Pull request merged**